ddl/column.go

```diff
@@ -1490,81 +1601,30 @@ func (w *worker) doModifyColumn(
 }
 
 func adjustColumnInfoInModifyColumn(
-	job *model.Job, tblInfo *model.TableInfo, newCol, oldCol *model.ColumnInfo, pos *ast.ColumnPosition, changingColUniqueLowerName string) error {
+	job *model.Job, tblInfo *model.TableInfo, newCol, oldCol *model.ColumnInfo, pos *ast.ColumnPosition) error {
 	// We need the latest column's offset and state. This information can be obtained from the store.
 	newCol.Offset = oldCol.Offset
 	newCol.State = oldCol.State
-	// Calculate column's new position.
-	oldPos, newPos := oldCol.Offset, oldCol.Offset
-	if pos.Tp == ast.ColumnPositionAfter {
-		// TODO: The check of "RelativeColumn" can be checked in advance. When "EnableChangeColumnType" is true, unnecessary state changes can be reduced.
-		if oldCol.Name.L == pos.RelativeColumn.Name.L {
-			// `alter table tableName modify column b int after b` will return ver,ErrColumnNotExists.
-			// Modified the type definition of 'null' to 'not null' before this, so rollback the job when an error occurs.
-			job.State = model.JobStateRollingback
-			return infoschema.ErrColumnNotExists.GenWithStackByArgs(oldCol.Name, tblInfo.Name)
-		}
-
-		relative := model.FindColumnInfo(tblInfo.Columns, pos.RelativeColumn.Name.L)
-		if relative == nil || relative.State != model.StatePublic {
-			job.State = model.JobStateRollingback
-			return infoschema.ErrColumnNotExists.GenWithStackByArgs(pos.RelativeColumn, tblInfo.Name)
-		}
-
-		if relative.Offset < oldPos {
-			newPos = relative.Offset + 1
-		} else {
-			newPos = relative.Offset
-		}
-	} else if pos.Tp == ast.ColumnPositionFirst {
-		newPos = 0
+	if pos != nil && pos.RelativeColumn != nil && oldCol.Name.L == pos.RelativeColumn.Name.L {
+		// `alter table tableName modify column b int after b` will return ver, ErrColumnNotExists.
+		// Modified the type definition of 'null' to 'not null' before this, so rollback the job when an error occurs.
+		job.State = model.JobStateRollingback
+		return infoschema.ErrColumnNotExists.GenWithStackByArgs(oldCol.Name, tblInfo.Name)
 	}
-
-	columnChanged := make(map[string]*model.ColumnInfo)
-	columnChanged[oldCol.Name.L] = newCol
-
-	if newPos == oldPos {
-		tblInfo.Columns[newPos] = newCol
-	} else {
-		cols := tblInfo.Columns
-
-		// Reorder columns in place.
-		if newPos < oldPos {
-			// ******** +(new) ****** -(old) ********
-			// [newPos:old-1] should shift one step to the right.
-			copy(cols[newPos+1:], cols[newPos:oldPos])
-		} else {
-			// ******** -(old) ****** +(new) ********
-			// [old+1:newPos] should shift one step to the left.
-			copy(cols[oldPos:], cols[oldPos+1:newPos+1])
-		}
-		cols[newPos] = newCol
-
-		for i, col := range tblInfo.Columns {
-			if col.Offset != i {
-				columnChanged[col.Name.L] = col
-				col.Offset = i
-			}
-		}
+	offset, err := locateOffsetForColumn(pos, tblInfo)
```

Comment:
Could we add an argument like `currentOffset`(`add column` can pass `len(cols)-1`)？ Then we can remove the judgment：https://github.com/pingcap/tidb/pull/33631/files#diff-42e0180cdc8c14def674a9e45d908d6700f6a8ccba0f9ff4b14b27c0daa49859R1620.